[Fetch API] Headers iteration should not happen on cached headers list

youennf opened the pull request:
#### 68f5a2f24c94431e60a9517543ad28b25e5809c3
<pre>
[Fetch API] Headers iteration should not happen on cached headers list
<a href="https://bugs.webkit.org/show_bug.cgi?id=246526">https://bugs.webkit.org/show_bug.cgi?id=246526</a>
rdar://problem/101235580

Reviewed by Chris Dumez.

We introduce a change counter that is incremented anytime a JS API is mutating the headers (append, delete, set).
We recreate the header list as needed, either on the first time or if the change counter tells us to do so.

Covered by resynced WPT tests.

* LayoutTests/imported/w3c/web-platform-tests/fetch/api/headers/headers-basic.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/headers/headers-basic.any.js:
(test):
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/headers/headers-basic.any.worker-expected.txt:
* Source/WebCore/Modules/fetch/FetchHeaders.cpp:
(WebCore::FetchHeaders::append):
(WebCore::FetchHeaders::remove):
(WebCore::FetchHeaders::set):
(WebCore::FetchHeaders::Iterator::next):
(WebCore::FetchHeaders::Iterator::Iterator):
* Source/WebCore/Modules/fetch/FetchHeaders.h:

Canonical link: <a href="https://commits.webkit.org/255639@main">https://commits.webkit.org/255639@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/80778856c2941df68b9fe32fb3c8cc80f2a2db70

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93098 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2298 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23746 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102805 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/163114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97100 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2306 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30626 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85488 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98923 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98763 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1590 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79572 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28511 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83504 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/83269 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71620 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37019 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17154 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34831 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18388 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3907 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38702 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40929 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40632 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37592 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/fetch/api/headers/headers-basic.any-expected.txt:
```diff
@@ -18,4 +18,8 @@ PASS Check values method
 PASS Check entries method
 PASS Check Symbol.iterator method
 PASS Check forEach method
+PASS Iteration skips elements removed while iterating
```
How are Gecko and Blink doing on those new checks?

Also, is this behavior specified anywhere? It is not obvious to me that the new behavior is better so I am curious.

youennf replied:
Right, as pointed by @annevk, we might want to check other iterators.
Maybe we should make it more explicit in how our iterators work as well.

cdumez reviewed (approved):
I see now that the WPT updates were not written by us and that this is merely a re-sync.

LGTM.

youennf committed:
[Fetch API] Headers iteration should not happen on cached headers list
https://bugs.webkit.org/show_bug.cgi?id=246526
rdar://problem/101235580

Reviewed by Chris Dumez.

We introduce a change counter that is incremented anytime a JS API is mutating the headers (append, delete, set).
We recreate the header list as needed, either on the first time or if the change counter tells us to do so.

Covered by resynced WPT tests.

* LayoutTests/imported/w3c/web-platform-tests/fetch/api/headers/headers-basic.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/headers/headers-basic.any.js:
(test):
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/headers/headers-basic.any.worker-expected.txt:
* Source/WebCore/Modules/fetch/FetchHeaders.cpp:
(WebCore::FetchHeaders::append):
(WebCore::FetchHeaders::remove):
(WebCore::FetchHeaders::set):
(WebCore::FetchHeaders::Iterator::next):
(WebCore::FetchHeaders::Iterator::Iterator):
* Source/WebCore/Modules/fetch/FetchHeaders.h:

Canonical link: https://commits.webkit.org/255639@main
